### PR TITLE
Add dummy Cuckoo API

### PR DESCRIPTION
**Pull request by Jack28**
Add initial dummy_cuckoo_api for testing
    
To effectively test API mode we need a counter part.
This is a first and very rudimentary implementation of the Cuckoo API.
Allows testing the API link to Cuckoo Sandbox
    
It allows to add jobs, view a jobs state, fetch reports and check status.

### CODE REVIEW

File: bin/dummy_cuckoo_api.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    """ Intermediate class to extend normal Resource by Queue """
+    def __init__(self, queue):
+        self.queue = queue
+
+
+class Status(QueueResource):
+    """ Resource to get Dummy Cuckoo's status on tasks """
+    def get(self):
+        """ Returns number of tasks in the respective state """
+        result = {'tasks':
```
Comment (michaelweiser) [inline bin/dummy_cuckoo_api.py]: pycodestyle says:
```
bin/dummy_cuckoo_api.py:64:23: E127 continuation line over-indented for visual indent
bin/dummy_cuckoo_api.py:65:80: E501 line too long (83 > 79 characters)
bin/dummy_cuckoo_api.py:70:18: E124 closing bracket does not match visual indentation
```
That should be easily fixed:
```
@@ -60,14 +60,14 @@ class Status(QueueResource):
     """ Resource to get Dummy Cuckoo's status on tasks """
     def get(self):
         """ Returns number of tasks in the respective state """
-        result = {'tasks':
-                      {
-                          'reported': self.queue.nextid - 1 - len(self.queue.jobs),
-                          'running': len(self.queue.jobs),
-                          'total': self.queue.nextid - 1,
-                          'dummyAPIqueue': self.queue.jobs
-                      }
-                 }
+        result = {
+            'tasks': {
+                'reported': self.queue.nextid - 1 - len(self.queue.jobs),
+                'running': len(self.queue.jobs),
+                'total': self.queue.nextid - 1,
+                'dummyAPIqueue': self.queue.jobs,
+            }
+        }
         return flask.jsonify(result)

```
Also note the comma on `dummyAPIqueue` for whenever we extend or reorder that list.

File: bin/dummy_cuckoo_api.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        except FileNotFoundError:
+            report = {
+                "info": {
+                    "score": 10,
+                    "id": job_id,
+                },
+                "signatures": [
+                    {
+                        "description":
+                            "Malicious document featuring Office DDE has been identified",
```
Comment (michaelweiser) [inline bin/dummy_cuckoo_api.py]: Brackets and string constant concatenation can be used here to break this long line.
``` python
>>> ("aa "
... "bb")
'aa bb'
```
Note the missing comma within the round brackets:
```
                "signatures": [
                    {
                        "description": (
                            "Malicious document featuring Office "
                            "DDE has been identified"
                        ),
                    }
                ],
```

File: bin/dummy_cuckoo_api.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        else:
+            print("Job_id {} not found".format(job_id))
+            abort(404, message="Job_id {} doesn't exist".format(job_id))
+        return flask.jsonify(result)
+
+
+class Report(QueueResource):
+    """ Resource to accessed finished reports """
+    def get(self, job_id):
+        """ Unless Cuckoo report placed in storage/analyses/0/reports/report.json
```
Comment (michaelweiser) [inline bin/dummy_cuckoo_api.py]: ... report *is* placed ...?

Do these lines need to be so long?
```
@@ -97,10 +97,11 @@ class View(QueueResource):
 class Report(QueueResource):
     """ Resource to accessed finished reports """
     def get(self, job_id):
-        """ Unless Cuckoo report placed in storage/analyses/0/reports/report.json
-        returns minimal report """
+        """ Unless Cuckoo report placed in
+        storage/analyses/0/reports/report.json returns minimal report """
         try:
-            with open('storage/analyses/0/reports/report.json', 'r') as content_file:
+            filename = 'storage/analyses/0/reports/report.json'
+            with open(filename, 'r') as content_file:
                 report = json.load(content_file)
         except FileNotFoundError:
             report = {
```

File: bin/dummy_cuckoo_api.py [hunk 4]
```diff
@@ -108,18 +126,4 @@ def post(self):
 
 
 if __name__ == '__main__':
-    PARSER = reqparse.RequestParser()
```
Comment (michaelweiser) [inline bin/dummy_cuckoo_api.py]: Was reqparse unnecessary or where did it go?

**Comment by michaelweiser** [comment]
This has no impact on the rest of the code and adds a valuable testing facility. So it can go into 2.0.

File: bin/dummy_cuckoo_api.py (hunk 4 above)
Comment (Jack28) [reply bin/dummy_cuckoo_api.py]: It was a left over

Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Good riddance. Cheers. :)

File: bin/dummy_cuckoo_api.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+# pip install flask flask-restful """
+
+import json
+import random
+import sys
+import flask
+from flask_restful import abort, Resource, Api
+
+
+def run():
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits.

File: bin/dummy_cuckoo_api.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    """ Run the API """
+
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits.

File: bin/dummy_cuckoo_api.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            # job is running
+            if int(job_id) in self.queue.jobs:
+                print("Job_id {} running".format(job_id))
+                result = {'task': {'status': 'running'}}
+                # job is done
+                if random.randint(0, 2) == 0:
+                    print("Job_id {} reported".format(job_id))
+                    result['task']['status'] = "reported"
+                    self.queue.delete_job(int(job_id))
+        else:
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits.

File: bin/dummy_cuckoo_api.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    """ Resource to access state of task """
+    def get(self, job_id):
+        """ Returns the current status of a given job_id.
+        Jobs finish by asking its state with a probability of 1/3 """
+        # job exists
+        if int(job_id) < self.queue.nextid:
+            result = {'task': {'status': 'reported'}}
+            # job is running
+            if int(job_id) in self.queue.jobs:
+                print("Job_id {} running".format(job_id))
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits.

File: bin/dummy_cuckoo_api.py (hunk 1 above)
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits.

File: bin/dummy_cuckoo_api.py (hunk 3 above)
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits.

File: bin/dummy_cuckoo_api.py (hunk 2 above)
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits. Didn't know that you don't even need the brackets in a dict. Net.

File: bin/dummy_cuckoo_api.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        return flask.jsonify(result)
+
+
+class View(QueueResource):
+    """ Resource to access state of task """
+    def get(self, job_id):
+        """ Returns the current status of a given job_id.
+        Jobs finish by asking its state with a probability of 1/3 """
+        # job exists
+        if int(job_id) < self.queue.nextid:
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits.

File: bin/dummy_cuckoo_api.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+Allows testing the API link to Cuckoo Sandbox
+
+# requires
+# pip install flask flask-restful """
+
+import json
+import random
+import sys
+import flask
+from flask_restful import abort, Resource, Api
```
[the remark this answers is not in the document]
Comment (michaelweiser) [reply bin/dummy_cuckoo_api.py]: Addressed in new commits.

**Comment by michaelweiser** [comment]
Awesome!

**Pull request merged**